server/src/main/scala/org/http4s/armeria/server/ArmeriaHttp4sHandler.scala

```diff
@@ -22,43 +26,48 @@ import com.linecorp.armeria.common.{
 }
 import com.linecorp.armeria.common.util.Version
 import com.linecorp.armeria.server.{HttpService, ServiceRequestContext}
-import io.chrisdavenport.vault.{Key => VaultKey, Vault}
+import org.typelevel.vault.{Key => VaultKey, Vault}
 import fs2._
 import fs2.interop.reactivestreams._
 import java.net.InetSocketAddress
+
 import org.http4s.internal.CollectionCompat.CollectionConverters._
 import ArmeriaHttp4sHandler.{RightUnit, canHasBody, defaultVault, toHttp4sMethod}
+import cats.effect.std.Dispatcher
+import com.comcast.ip4s.SocketAddress
 import org.http4s.server.{
   DefaultServiceErrorHandler,
   SecureSession,
   ServerRequestKeys,
   ServiceErrorHandler
 }
+import org.typelevel.ci.CIString
 import scodec.bits.ByteVector
 
 /** An [[HttpService]] that handles the specified [[HttpApp]] under the specified `prefix`. */
 private[armeria] class ArmeriaHttp4sHandler[F[_]](
     prefix: String,
     service: HttpApp[F],
-    serviceErrorHandler: ServiceErrorHandler[F]
-)(implicit F: ConcurrentEffect[F])
+    serviceErrorHandler: ServiceErrorHandler[F],
+    dispatcher: Dispatcher[F]
+)(implicit F: Async[F])
     extends HttpService {
 
-  val prefixLength: Int = if (prefix.endsWith("/")) prefix.length - 1 else prefix.length
+  val prefixLength: Int = Uri.Path.unsafeFromString(prefix).segments.size
   // micro-optimization: unwrap the service and call its .run directly
   private val serviceFn: Request[F] => F[Response[F]] = service.run
 
   override def serve(ctx: ServiceRequestContext, req: HttpRequest): HttpResponse = {
     val responseWriter = HttpResponse.streaming()
-    F.runAsync(
+    dispatcher.unsafeRunAndForget(
       toRequest(ctx, req)
-        .fold(onParseFailure(_, responseWriter), handleRequest(_, responseWriter))) {
-      case Right(_) =>
-        IO.unit
-      case Left(ex) =>
-        discardReturn(responseWriter.close(ex))
-        IO.unit
-    }.unsafeRunSync()
+        .fold(onParseFailure(_, responseWriter), handleRequest(_, responseWriter))
+        .handleErrorWith { ex =>
+          F.delay {
+            discardReturn(responseWriter.close(ex))
+          }
+        }
```

Comment:
Could use just use `handleError`?
```suggestion
        .handleError { ex =>
            discardReturn(responseWriter.close(ex))
        }
```

Reply:
My logic was that `responseWriter.close` is an effectful operation that should be encapsulated within the `F`. But I guess this whole function is "unsafe" anyway.